main.py

```diff
@@ -92,78 +89,99 @@ def format_ids(options_string):
     interval = "1D"
     interval_percent = "24h_change_percent"
 
-for user in users_data:
-    if "unsubscribe?" in user:
-        continue
-    user_email = user["emailAddress"]
-    user_options = user.get("anyExtraDataYou'dLikeInYourReport?", None)
-    print(f"Compiling daily report for {user_email}...")
-
-    msg = MIMEMultipart()
-    msg["From"] = email_man.smtp_username
-    msg["To"] = user_email
-    msg['Subject'] = f"BTC {close_significance} Close: {crypto_data['bitcoin']['price']}"
-
-    message_body = "<p>"
-    message_body += default_msg(
-        ticker=crypto_data["bitcoin"]["ticker_upper"],
-        price=crypto_data["bitcoin"]["price"],
-        percent_change=crypto_data["bitcoin"]["24h_change_percent"]
-    )
-    if close_significance == MONTHLY or close_significance == WEEKLY:
-        btc_wm_up_down = up_down_icon(crypto_data["bitcoin"][interval_percent])
-        message_body += (f"{close_significance.title()} Change ({interval}) "
-                         f"{btc_wm_up_down} {crypto_data['bitcoin'][interval_percent]}%")
-    message_body += "</p><p>"
-
-    if user_options:
-        user_choices = format_ids(user_options)[::2]
-        message_body += "Crypto:<br>"
-        for choice in user_choices:
-            crypto_dict = crypto_data[choice.lower()]
-            message_body += default_msg(
-                ticker=crypto_dict["ticker_upper"],
-                price=crypto_dict["price"],
-                percent_change=crypto_dict["24h_change_percent"]
-            )
-            if close_significance == MONTHLY or close_significance == WEEKLY:
-                message_body += htf_msg(
-                    timeframe=interval,
-                    percent_change=crypto_dict[interval_percent]
-                )
-        message_body += "</p><p>"
+stock_market_open = True
+if day_of_week in ["0", "6"]:
+    stock_market_open = False
+
+
+async def main():
+
+    async with asyncio.TaskGroup() as tg:
+        tg.create_task(email_man.get_emails_data())
+        tg.create_task(crypto_man.get_global_crypto_data())
+        for asset in crypto_man.cg_assets_list:
+            tg.create_task(crypto_man.get_crypto_data(asset))
+        if stock_market_open:
+            for ind in stock_man.index_list:
+                tg.create_task(stock_man.get_index_data(ind))
+                await asyncio.sleep(0.1)  # Rate limiting
```

Comment:
what's the purpose of taskgroup in place of async/await? you're already using `gather`

Reply:
I read it handles errors better than straight up using gather because gather will still try to complete every other task even after one has failed (?) so didn't want to group everything into a single gather. But also the way I'm thinking about async right now is if I await something here that isn't a .create_task(), it will wait for that task to complete before moving onto the next await, yeah? I initlally drafted this portion of code with a "tasks" list that I'd append every create_task() to, and then gather(*tasks) at the end, but switched to TaskGroup(). lmk if I'm wrong or if there's a better way to do this